### PR TITLE
[LIVY-658]RSCDriver should catch exception if cancel job failed during shutdown

### PR DESCRIPTION
**Pull request by yantzu**
## What changes were proposed in this pull request?

Currently, if startup meet exception, exception will trigger spark to shutdown, then trigger cancel job, but cancel job will throw another exception due to spark is not initialized. The new exception will swallow the old exception.

https://issues.apache.org/jira/browse/LIVY-658

Before changes:
cancel job exception

After changes:
cancel job exception after fix

## How was this patch tested?

Tested manually， and  add unit test.

Please review https://livy.incubator.apache.org/community/ before opening a pull request.

### CODE REVIEW

**Comment by jerryshao** [comment]
CC @yiheng @runzhiwang please help to review.

**Comment by runzhiwang** [comment]
It's easy to find the exception was caused by canceling job whether catch or not ?

**Comment by yantzu** [comment]
Sorry @runzhiwang , I cannot catch your point, could you please explain more?

**Comment by runzhiwang** [comment]
@yantzu If do not catch the exception, it's also easy to find cancel job will cause exception. Is it necessary to catch ?

**Comment by yantzu** [comment]
Without catch cancel exception, the original exception("scala.refrect.internal.FactalError: object Predef does not have a member classOf" in the screen-shoot) is concealed by NullPointerException, it is not easy to find application error root cause(which is FactalError in this case). @runzhiwang

**Comment by runzhiwang** [comment]
@yantzu Is the original exception helpful to find the reason is fail to startup.

**Comment by yantzu** [comment]
original exception is the cause of application failure. @runzhiwang

**Comment by yiheng** [comment]
can we fix the annoying `NullPointerException`?

**Comment by yantzu** [comment]
The unit test I add demonstrate how NullPointerException happens: a request send to RSCDriver while it is initializing, after initialization failed, RSCDriver try to cancel job. In real case, it is beeline send a "use default" SQL to thrift server then RSCDriver. 
I think it is possible to avoid NPE in this case, but not sure if it is what we should do, because:
1, there may be other reason cause cancel job failure, to fix all those would be diffcult
2, catch and log exception when cancel job give RSCDriver a chance to cancel other job and shutdown context correctly
3, RSCDriver already catch shutdownContext() and shutdownServer() call

**Comment by jerryshao** [comment]
Retrigger the test.

File: rsc/src/test/java/org/apache/livy/rsc/driver/TestRSCDriver.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.livy.rsc.driver;
+
+import org.apache.spark.SparkConf;
+
```
Comment (jerryshao) [inline rsc/src/test/java/org/apache/livy/rsc/driver/TestRSCDriver.java]: Remove this blank line.

File: rsc/src/test/java/org/apache/livy/rsc/driver/TestRSCDriver.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import org.junit.Test;
+
+import org.apache.livy.rsc.BaseProtocol;
+import org.apache.livy.rsc.RSCConf;
+
+public class TestRSCDriver {
+  @Test(expected = IllegalArgumentException.class)
+  public void testCancelJobFailedAfterInitializeFailedShouldThrowOriginalException()
```
Comment (jerryshao) [inline rsc/src/test/java/org/apache/livy/rsc/driver/TestRSCDriver.java]: Can you please shorten the method name?

Comment (yantzu) [reply rsc/src/test/java/org/apache/livy/rsc/driver/TestRSCDriver.java]: fixed

**Comment by jerryshao** [comment]
LGTM, merging to master branch.

**Pull request closed without merging**